src/redisraft.c

```diff
@@ -370,6 +370,161 @@ static void handleInfoCommand(RedisRaftCtx *rr,
     RedisModule_FreeCallReply(reply);
 }
 
+typedef struct MigrationInfo {
+    char shard_group_id[RAFT_DBID_LEN+1];
+    char username[256];
+    char password[256];
+} MigrationInfo;
+
+static MigrationInfo * extractMigrationInfo(RedisRaftCtx *rr, RaftRedisCommand *cmd)
+{
+    MigrationInfo * ret = RedisModule_Calloc(1, sizeof(MigrationInfo));
+    size_t str_len;
+    const char * str = RedisModule_StringPtrLen(cmd->argv[1], &str_len);
+    if (str_len > RAFT_DBID_LEN) {
+        LOG_WARNING("shardgroup id is too long");
+        goto fail;
+    }
+    if (getShardGroupById(rr, str) == NULL) {
```

Comment:
don't understand the Q.  here we are validating the parameters.  so what has happened.

we've made ourselves migrating, and another shardgroup importing.  we give the migrate command the import shardgroup id.  we validate that at this point in time in state machine we know about this id.  if we don't, we fail early.  unsure there's anything really bad about that.

Reply:
Why do we have to lock? When we are about to apply, we can check this condition there and if something is wrong, we don't lock the keys. Again, here we are reading state machine out of raft logic, this check is not reliable or deterministic anyway.

Reply:
I dont understand the Q.  are you saying that just because the id exits here, it might not exist later?

yes.  that's true.  It can possibly have a false negative failure here.  I'm not sure why that would be a problem (as long as we also ensure its valid later and not rely on it being valid at this point in time).

If you really find this to be a code smell, I can refactor it.

or am I misunderstanding the Q?

Reply:
1- I'm just against reading state machine map directly, as always :) . `getShardGroupById()` does that internally, accesses state machine without going through log. So, I just wonder if there is a good reason to do it here. If this is an important check, we can also serialize `shard group id` and do that check in the `lockKeys()` as the first thing. That was my suggestion. 

This check here can return false negative and maybe false positive in rare conditions? A stale request can pass this check if the state machine is not advanced enough? So, is this going to be a problem then? If the answer is yes, we have to handle this inside state machine anyway. No? I just think that these kind of checks add complexity and they are harder to reason about. It'll also make debugging/testing harder as it relies on timing. So, just want to be sure we are doing the correct thing. 

2- Slightly related, I realized that we don't have any protection against stale migrate commands. Assume that client sends migrate command, RedisRaft completes the migration but client gets disconnected before getting the ack. Client sends the same command again and we lock the keys without checking them. 
- If the command locks non-existing keys, these keys will be leaked as they will be in the map forever.
- If the command locks the keys that we are not migrating, that's another problem I guess. 

So, do we need some sort of validation here? Slot validation and we need to check key existence? My assumption was it is always okay to lock the keys even a command fails to complete the operation because the whole operation will be retried and these keys will be moved eventually. So, I guess we need to make sure we lock the keys that we will actually move.

Reply:
I thought I posted twice yesterday, deleted one comment. Now I see, I deleted both probably or github is having hard time

Reply:
@sjpotter Yes, personally I like 3 better. Just to make everything deterministic again. Also, I realized something else, just take a look at question 2. Probably, we need some sort of validation there? Wdyt?

Reply:
yes, this code wasn't as designed.

```
    for (size_t i = 0; i < num_keys; i++) {
        size_t str_len;
        const char * str = RedisModule_StringPtrLen(keys[i], &str_len);
        LOG_WARNING("locking %.*s", (int) str_len, str);
        RedisModule_DictSet(rr->locked_keys, keys[i], NULL);
        RedisModule_FreeString(rr->ctx, keys[i]);
    }
```

changing to

```
    for (size_t i = 0; i < num_keys; i++) {
        RedisModuleString * key = RedisModule_StringPtrLen(keys[i];
        if (RedisModule_KeyExists(rr->ctx, key)) {
            size_t str_len;
            const char *str = RedisModule_StringPtrLen(key, &str_len);
            LOG_WARNING("locking %.*s", (int) str_len, str);
            RedisModule_DictSet(rr->locked_keys, keys[i], NULL);
        }
        RedisModule_FreeString(rr->ctx, key);
    }
```

Reply:
I have some more sanity checks I have to put into the state machine operation.  going over it now.